### PR TITLE
ODIN: Added identifier node to ast_node_t struct

### PR DESCRIPTION
**Pull request by emacdo12**
#### Description

Added that to structure and changed everything that was affected. Updated graphviz printing function.

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

File: ODIN_II/regression_test/benchmark/verilog/large/arm_core_take_2.v [hunk 1]
```diff
@@ -0,0 +1,185 @@
+
```
Comment (jeanlego) [inline ODIN_II/regression_test/benchmark/verilog/large/arm_core_take_2.v]: is this test still necessary?

File: ODIN_II/regression_test/benchmark/verilog/large/arm_core_copy.v [hunk 2]
```diff
@@ -0,0 +1,250 @@
+
```
Comment (jeanlego) [inline ODIN_II/regression_test/benchmark/verilog/large/arm_core_copy.v]: is this test still necessary?

File: ODIN_II/regression_test/benchmark/task/syntax/synthesis_result.json [hunk 3]
```diff
@@ -409,35 +409,23 @@
         "test_name": "syntax/bm_function_1/k6_frac_N10_frac_chain_mem32K_40nm",
         "architecture": "k6_frac_N10_frac_chain_mem32K_40nm.xml",
         "verilog": "bm_function_1.v",
-        "exit": 134,
-        "errors": [
-            "bm_function_1.v:40:1 [AST] Missing declaration of this symbol simple_func"
-        ]
+        "exit": 134
```
Comment (jeanlego) [inline ODIN_II/regression_test/benchmark/task/syntax/synthesis_result.json:412]: the error changed here, and the missing declaration didn't move to warnings
can you check this?

File: ODIN_II/regression_test/benchmark/task/keywords/defparam/simulation_result.json [hunk 4]
```diff
@@ -115,6 +115,53 @@
...
         "Total Node": 12
     },
+    "defparam/defparam_depth_2/no_arch": {
+        "test_name": "defparam/defparam_depth_2/no_arch",
+        "blif": "defparam_depth_2.blif",
+        "exit": 134,
+        "errors": [
+            "[OUTPUT_BLIF] Vector files differ."
+        ],
+        "warnings": [
```
Comment (jeanlego) [inline ODIN_II/regression_test/benchmark/task/keywords/defparam/simulation_result.json:125]: can you check these? it seems that synthesis passes, but it doesnt produce the right vector

File: ODIN_II/SRC/ast_elaborate.cpp [hunk 5]
```diff
@@ -2308,17 +2297,17 @@ void convert_2D_to_1D_array(ast_node_t** var_declare) {
...
+        error_message(AST, (*var_declare)->identifier_node->loc, "%s",
                       "Odin doesn't support negative number in index.");
     }
 
-    char* name = (*var_declare)->children[0]->types.identifier;
+    char* name = (*var_declare)->identifier_node->types.identifier;
 
     if (addr_min != 0 || addr_min1 != 0) {
-        error_message(AST, (*var_declare)->children[0]->loc,
+        error_message(AST, (*var_declare)->identifier_node->loc,
```
Comment (jeanlego) [inline ODIN_II/SRC/ast_elaborate.cpp]: just use (*var_declare)->loc

File: ODIN_II/SRC/ast_elaborate.cpp [hunk 6]
```diff
@@ -2353,17 +2342,17 @@ void convert_2D_to_1D_array_ref(ast_node_t** node, sc_hierarchy* local_ref) {
...
 
-    ast_node_t* array_row = (*node)->children[1];
-    ast_node_t* array_col = (*node)->children[2];
+    ast_node_t* array_row = (*node)->children[0];
+    ast_node_t* array_col = (*node)->children[1];
 
     // build the new AST
-    ast_node_t* new_node_1 = newBinaryOperation(MULTIPLY, array_row, array_size, (*node)->children[0]->loc);
-    ast_node_t* new_node_2 = newBinaryOperation(ADD, new_node_1, array_col, (*node)->children[0]->loc);
+    ast_node_t* new_node_1 = newBinaryOperation(MULTIPLY, array_row, array_size, (*node)->identifier_node->loc);
```
Comment (jeanlego) [inline ODIN_II/SRC/ast_elaborate.cpp]: replace all occurence of `some_node->identifier_node->loc` to `some_node->loc`

File: ODIN_II/SRC/netlist_create_from_ast.cpp [hunk 7]
```diff
@@ -552,14 +550,16 @@ signal_list_t* netlist_expand_ast_of_module(ast_node_t** node_ref, char* instanc
         if (skip_children == false) {
             /* traverse all the children */
             for (i = 0; i < node->num_children; i++) {
-                if (child_skip_list && child_skip_list[i] == false) {
-                    sc_hierarchy* child_ref = local_ref;
-                    if (node->children[i]->types.hierarchy != NULL) {
-                        child_ref = node->children[i]->types.hierarchy;
-                    }
+                if (node->children[i]) {
```
Comment (jeanlego) [inline ODIN_II/SRC/netlist_create_from_ast.cpp]: I agree that this should be checked, but I'm warry to add it wereas it wasnt there before and never cause any segmentation fault.
I think this will be better served by an `oassert`.
Was this added because some nodes got thrown as NULL when they werent before?

File: ODIN_II/SRC/netlist_create_from_ast.cpp [hunk 8]
```diff
@@ -873,15 +873,15 @@ void create_top_output_nodes(ast_node_t* module, char* instance_name_prefix, sc_
...
                     /* IF - this is a identifier then find it in the output_nets and hook it up to a newly created node */
                     /* get the name of the pin */
-                    full_name = make_full_ref_name(instance_name_prefix, NULL, NULL, var_declare->children[0]->types.identifier, -1);
+                    full_name = make_full_ref_name(instance_name_prefix, NULL, NULL, var_declare->identifier_node->types.identifier, -1);
 
                     /* check if the instantiation pin exists. */
                     if ((sc_spot = sc_lookup_string(output_nets_sc, full_name)) == -1) {
-                        error_message(NETLIST, var_declare->children[0]->loc,
-                                      "Output pin (%s) is not hooked up!!!\n", full_name);
+                        warning_message(NETLIST, var_declare->identifier_node->loc,
```
Comment (jeanlego) [inline ODIN_II/SRC/netlist_create_from_ast.cpp]: see comment about using `identifier_node->loc` above

File: ODIN_II/SRC/netlist_create_from_ast.cpp [hunk 9]
```diff
@@ -1436,18 +1436,18 @@ void connect_memory_and_alias(ast_node_t* hb_instance, char* instance_name_prefi
...
 
             /* this will crash because of an oassert. we keep it but we work around the problem here */
-            if (hb_instance_var_node->type == RANGE_REF) {
-                full_name = make_full_ref_name(instance_name_prefix, NULL, NULL, NULL, -1);
-            } else {
-                name_of_hb_input = get_name_of_pin_at_bit(hb_instance_var_node, -1, instance_name_prefix, local_ref);
-                full_name = make_full_ref_name(instance_name_prefix, NULL, NULL, name_of_hb_input, -1);
-                vtr::free(name_of_hb_input);
-            }
+            // if (hb_instance_var_node->type == RANGE_REF) {
```
Comment (jeanlego) [inline ODIN_II/SRC/netlist_create_from_ast.cpp]: can you uncomment for now, It should be tested in a separate PR with tests and all (sorry forgot to remove my changes.)

File: ODIN_II/SRC/netlist_create_from_ast.cpp [hunk 10]
```diff
@@ -1467,24 +1467,24 @@ void connect_memory_and_alias(ast_node_t* hb_instance, char* instance_name_prefi
...
+                                                    hb_connect_list->children[i]->identifier_node->types.identifier, j);
                 } else {
                     oassert(j == 0);
                     /* this will crash because of an oassert. we keep it but we work around the problem here */
                     if (hb_instance_var_node->type == RANGE_REF) {
                         full_name = make_full_ref_name(instance_name_prefix, NULL, NULL, NULL, -1);
                     } else {
                         name_of_hb_input = get_name_of_pin_at_bit(hb_instance_var_node, -1, instance_name_prefix, local_ref);
-                        full_name = make_full_ref_name(instance_name_prefix, NULL, NULL, name_of_hb_input, j);
+                        full_name = make_full_ref_name(instance_name_prefix, NULL, NULL, name_of_hb_input, -1);
```
Comment (jeanlego) [inline ODIN_II/SRC/netlist_create_from_ast.cpp]: idem. remove this change

File: ODIN_II/SRC/parse_making_ast.cpp [hunk 11]
```diff
@@ -339,17 +339,17 @@ ast_node_t* resolve_ports(ids top_type, ast_node_t* symbol_list) {
...
+                    if (symbol_list->children[j]->children[0] == NULL) {
+                        symbol_list->children[j]->children[0] = ast_node_deep_copy(this_port->children[0]);
                         symbol_list->children[j]->children[1] = ast_node_deep_copy(this_port->children[1]);
                         symbol_list->children[j]->children[2] = ast_node_deep_copy(this_port->children[2]);
                         symbol_list->children[j]->children[3] = ast_node_deep_copy(this_port->children[3]);
-                        symbol_list->children[j]->children[4] = ast_node_deep_copy(this_port->children[4]);
 
-                        if (this_port->num_children == 8) {
+                        if (this_port->num_children == 7) {
                             symbol_list->children[j]->children = (ast_node_t**)realloc(symbol_list->children[j]->children, sizeof(ast_node_t*) * 8);
```
Comment (jeanlego) [inline ODIN_II/SRC/parse_making_ast.cpp]: this should be `sizeof(ast_node_t*) * 7`

File: ODIN_II/SRC/parse_making_ast.cpp [hunk 12]
```diff
@@ -1783,15 +1818,15 @@ void graphVizOutputAst(std::string path, ast_node_t* top) {
     static int file_num = 0;
 
     /* open the file */
-    odin_sprintf(path_and_file, "%s/%s_ast.dot", path.c_str(), top->children[0]->types.identifier);
+    odin_sprintf(path_and_file, "%s/%s_ast.dot", path.c_str(), top->identifier_node->types.identifier);
     fp = fopen(path_and_file, "w");
     file_num++;
 
     /* open graph */
     fprintf(fp, "digraph G {\t\nranksep=.25;\n");
```
Comment (jeanlego) [inline ODIN_II/SRC/parse_making_ast.cpp]: while you are changing it, would you mind modifying this to 
```
fprintf(fp, "digraph G {\t\nrankdir=LR;\t\nranksep=.25;\n");
```
this will print the graph left to right, making it much more readable

File: ODIN_II/SRC/parse_making_ast.cpp [hunk 13]
```diff
@@ -1801,55 +1836,65 @@ void graphVizOutputAst(std::string path, ast_node_t* top) {
 /*---------------------------------------------------------------------------------------------
  * (function: graphVizOutputAst_traverse_node)
  *-------------------------------------------------------------------------------------------*/
-void graphVizOutputAst_traverse_node(FILE* fp, ast_node_t* node, ast_node_t* from, int from_num) {
+void graphVizOutputAst_traverse_node(FILE* fp, ast_node_t* node, ast_node_t* from, int from_num, bool identifier) {
```
Comment (jeanlego) [inline ODIN_II/SRC/parse_making_ast.cpp]: I think what you have done offers us a really good opportunity to combine identifiers inside a node, hopefully reducing the graph size and improve the readability

I think we should print the identifier as part of the node rather than separately.
To have it pretty print you need to use html tags

so the end-goal should print something along these line

```
[label=<identifier<br/>NODE_TYPE>]
```

and it will render roughly like this

```
,--------------------,
|     identifier     |
|    NODE_TYPE       |
'--------------------'
```

to do so you need to first print `fprintf(fp, "\t%d [label=<", my_label);` first

then 

```
 if (identifier) {
        fprintf(fp, "%s<br/>", node->types.identifier);
 }
```

then do the regular switch

then finish with

```
fprintf(fp, ">];\n"); 
```

**Review by jeanlego — CHANGES_REQUESTED** [review]
I've left some comment for you to address, I have some more reviews to do before this can be merged, but its looking pretty good :)

File: ODIN_II/regression_test/benchmark/task/syntax/synthesis_result.json (hunk 3 above)
Comment (emacdo12) [reply ODIN_II/regression_test/benchmark/task/syntax/synthesis_result.json:412]: Here there actually isn't a missing declaration, if you look at the benchmark it's there so its good that it's now passing. It fails when it tries to pass a_in[`BITS-1:0] because the function it's in assumes they're both numbers and doesn't do a binary op check. I mentioned that this bm changed. I'm double checking tho

File: ODIN_II/regression_test/benchmark/task/keywords/defparam/simulation_result.json (hunk 4 above)
Comment (emacdo12) [reply ODIN_II/regression_test/benchmark/task/keywords/defparam/simulation_result.json:125]: I think with some of the changes we made it passes synthesis now. The issue is if you look at the benchmark the defparams are in a seperate module which isn't supported. The change isn't being taken into consideration thus what's supposed to happen isn't which leads to the vector mismatch.

File: ODIN_II/SRC/parse_making_ast.cpp (hunk 12 above)
Comment (emacdo12) [reply ODIN_II/SRC/parse_making_ast.cpp]: The line under open graph right?

Comment (jeanlego) [reply ODIN_II/SRC/parse_making_ast.cpp]: Yep

File: ODIN_II/SRC/netlist_create_from_ast.cpp (hunk 7 above)
Comment (emacdo12) [reply ODIN_II/SRC/netlist_create_from_ast.cpp]: I think this was a quick fix on my part because something NULL was being passed in, in the new commit I added an oassert and everything is fine now. My guess is I corrected what was actual wrong with another benchmark

**Comment by jeanlego** [comment]
this looks good, only thing missing would be a smaller version of the arm_core problem you ran onto
make sure to comment the veriog about what its trying to test

**Comment by jeanlego** [comment]
add the test, squash the commits into one, and you can move this PR onto the main repo.
feel free to close this PR too.

thanks for this :)

**Comment by emacdo12** [comment]
> this looks good, only thing missing would be a smaller version of the arm_core problem you ran onto
> make sure to comment the veriog about what its trying to test

Right I'll add that, forgot!

File: ODIN_II/regression_test/benchmark/verilog/syntax/bm_unused_input_pins.v [hunk 14]
```diff
@@ -0,0 +1,19 @@
+/* This test checks that when you pass an input into a function and there are unused Odin doesn't fail but throws a warning */
```
Comment (jeanlego) [inline ODIN_II/regression_test/benchmark/verilog/syntax/bm_unused_input_pins.v:1]: looks good and the test is right! :)

**Comment by jeanlego** [comment]
that ready to go :)
